### ISSUE TITLE
MBID recording merges are not applied

### ISSUE THREAD

**Issue by phw** [acoustid/acoustid-server#114]
It looks like since a while merged recordings in MB don't get followed up on the AcoustID side. There are now many instances of unresolved merges. Some examples:

https://acoustid.org/track/b1eb0783-f5f3-4428-9733-2fbfd8e30d93
https://acoustid.org/track/91684050-2987-48e9-807a-4f7354305511
https://acoustid.org/track/1cc7cd8a-50c7-4951-a8fa-557c6b2b41e6

Maybe there is some issue with running the merge code from https://github.com/acoustid/acoustid-server/blob/5c05072a01571b6a13f971bfcac2184936967549/acoustid/data/track.py#L149 . 

See also the discussion at https://community.metabrainz.org/t/why-does-acoustid-sometimes-not-show-merged-recordings/659454

**Comment by Echelon666** [comment]
Thousands of people are waiting for it.

**Comment by Echelon666** [comment]
See also the discussion at https://community.metabrainz.org/t/worse-results-in-picard-2-10/663668

**Comment by Echelon666** [comment]
See also the discussion at https://community.metabrainz.org/t/picard-scan-still-returning-new-acoustid-submissions-as-standalone-recordings/603676

**Comment by jesus2099** [comment]
After recording merge:

- The AcoustID page will show an italic link with the old MBID, alongside the merged recording with its metadata, instead of merging old and merged MBID metadata
- Merged MB recording Fingerprints tab will not show fingerprints that used to be linked to old MBID, only target’s, instead of showing them all

Is it only an AcoustID problem or an MB plus AcoustID problem? 

See [Italic AcoustID](https://community.metabrainz.org/t/italic-acoustid/734798?u=jesus2099) topic.

**Comment by lalinsky** [comment]
I've re-enabled this process a few days ago.

https://community.metabrainz.org/t/merged-mb-recordings-are-now-being-merged-in-acoustid-again/745357

**Comment by emkey08** [comment]
Thanks, lalinsky.

Shouldn't the `merge_mbids` function also update the `track_mbid.updated` field within [this UPDATE statement](https://github.com/acoustid/acoustid-server/blob/b40d294422843137d9d6d0079f48e27bafd449fa/acoustid/data/track.py#L145), so that the merged MBID's will be replicated to https://data.acoustid.org?

**Comment by lalinsky** [comment]
@emkey08 That's a good catch, it should. So I'll have to rebuild the data.acoustid.org files.

Hm, it's actually even more complicated, because it also deletes track_mbid and there is no way to replicate that currently.

**Comment by emkey08** [comment]
I think a full rebuild of the data files isn't actually necessary, as it should also be OK to just bump `track_mbid.updated` of the affected rows now, so that the updates will eventually get written to the next round of daily replication files.

Regarding the deleted rows: could this also be handled by simply setting `track_mbid.disabled` instead of doing a physical delete? If not, then maybe consider adding a `track_mbid.deleted` field which indicates logical deletion.

**Comment by lalinsky** [comment]
The problem is, I don't know which rows were effected.

**Comment by emkey08** [comment]
Could we possibly use the following query to find all affected rows and flag them as updated? (This assumes that `mb.recording` and `mb.recording_gid_redirect` are up-to-date tables of the MusicBrainz database.)

```
UPDATE track_mbid SET updated = now() WHERE mbid IN (SELECT gid FROM mb.recording WHERE id IN (SELECT new_id FROM mb.recording_gid_redirect));
```

**Comment by emkey08** [comment]
Regarding replication of deleted records, we could possibly add the following table:

```
CREATE TABLE track_mbid_delete (
    track_mbid_id INTEGER NOT NULL,
    created TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP NOT NULL
)
```

and initialize it with the result of the following query:

```
SELECT generate_series(1, (SELECT max(id) FROM track_mbid)) EXCEPT SELECT id FROM track_mbid;
```

Then, when replicating this new `track_mbid_delete` table, the replicating clients could perform corresponding `DELETE` statements on the `track_mbid` table (but only **after** processing the `track_mbid-update` file of that same day, otherwise data integrity won't be guaranteed).

Of course, inserting into the new `track_mbid_delete` table would also have to happen [here](https://github.com/acoustid/acoustid-server/blob/b40d294422843137d9d6d0079f48e27bafd449fa/acoustid/data/track.py#L136).

**Comment by lalinsky** [comment]
> Could we possibly use the following query to find all affected rows and flag them as updated? (This assumes that `mb.recording` and `mb.recording_gid_redirect` are up-to-date tables of the MusicBrainz database.)
> 
> ```
> UPDATE track_mbid SET updated = now() WHERE mbid IN (SELECT gid FROM mb.recording WHERE id IN (SELECT new_id FROM mb.recording_gid_redirect));
> ```

Yes, this is what I'll end up doing, however it's not that simple, because I don't have both MusicBrainz DB and the fingerprint DB in a single PostgreSQL instance, so the join has to be done externally.

**Comment by emkey08** [comment]
After thinking again about this, it might actually be possible to solve this with neither a schema nor replication change.

Instead of doing a physical delete, the `merge_mbids` function could set `track_mbid.submission_count = 0` [here](https://github.com/acoustid/acoustid-server/blob/b40d294422843137d9d6d0079f48e27bafd449fa/acoustid/data/track.py#L139) to indicate a pending delete. Then, the replication code would first create the `track_mbid-update.jsonl.gz` file as usual, but only then execute the physical delete by running `DELETE FROM track_mbid WHERE submission_count = 0`.

Replication clients would simply also run that `DELETE` statement after applying a `track_mbid-update.jsonl.gz` file.

To correctly replicate the already executed `UPDATE` and `DELETE` commands in hindsight, you could also rebuild a temporary `track_mbid_replicated` table from all the currently existing `track_mbid-update.jsonl.gz` replication files, and then run the following queries (untested, please verify):

```
-- Flag all track_mbid records as updated which are different from the track_mbid_replicated version
UPDATE track_mbid AS tm SET updated = now() WHERE (SELECT count(*) = 0 FROM track_mbid_replicated WHERE id = tm.id AND mbid = tm.mbid AND submission_count = tm.submission_count AND disabled = tm.disabled);

-- Re-insert the deleted rows with submission_count = 0 and updated = now()
INSERT INTO track_mbid (id, track_id, mbid, submission_count, disabled, created, updated) SELECT id, track_id, mbid, 0, disabled, now() FROM track_mbid_replicated WHERE id NOT IN (SELECT id FROM track_mbid);
```

**Comment by emkey08** [comment]
> keep the old track_mbid row with the merged mbid there, but change sources to 0 and make it disabled.

Sounds good! Just to mention it, the `track_mbid.mbid` field should then of course not be updated [here](https://github.com/acoustid/acoustid-server/blob/b40d294422843137d9d6d0079f48e27bafd449fa/acoustid/data/track.py#L147) any longer, otherwise we could end up with two identical `(track_id, mbid)` records. Instead, `merge_mbids` must either update or insert the `track_mbid.mbid = target_mbid` record.

**Comment by lalinsky** [comment]
Yeah, exactly, that's the part that needs to be handled. Duplicate `(track_id, mbid)` pairs are prohibited by an unique constraint in PostgreSQL, so it would fail. Yes, the point is to never update the `mbid` field of `track_mbid`, only insert new records and disable the old ones.

**Comment by emkey08** [comment]
Hm, I fear that the existing replication is actually broken by now. Some of the already merged `track_mbid` records have likely been updated again since last week, and hence some of the merged `track_mbid` records will actually have made it to the replication. This means that the replication data feed potentially contains duplicate `(track_id, mbid)` pairs now.

In order to avoid a full rebuild of all the data files, would it also be possible to just wipe the `track_mbid-update.jsonl.gz` files since `2025-02-21` (if that's the correct date where merging began), then fix the `track_mbid` table (if possible), and then simply dump the entire table within a follow-up `track_mbid-update.jsonl.gz` file?

**Comment by emkey08** [comment]
Here is a possible approach to fix the `track_mbid` table:

1) Re-assemble a `track_mbid_replica` table as per snapshot `2024-02-20` (i.e. before the merge) by using the existing `track_mbid-update.jsonl.gz` data files.

2) Re-apply merging to the entire `track_mbid_replica` table by using the new fixed merge algorithm. The replica will then have correct `updated` fields.

3) Re-apply all updates to the `track_mbid_replica` table which happened after `2024-02-20`:

```
INSERT INTO track_mbid_replica (track_id, mbid, submission_count, disabled, created, updated)
    SELECT (track_id, mbid, submission_count, disabled, created, updated)
        FROM track_mbid
        WHERE created >= '2024-02-21' OR updated >= '2024-02-21'
    ON CONFLICT (track_id, mbid) DO UPDATE SET
        submission_count = EXCLUDED.submission_count,
        disabled = EXCLUDED.disabled,
        updated = EXCLUDED.updated;
```

4) Verify that the data in `track_mbid_replica` is correct by sanity checking it against the current `track_mbid`.

5) Rename `track_mbid` table to `track_mbid_backup`.

6) Rename `track_mbid_replica` table to `track_mbid`.

7) Empty the `track_mbid-update.jsonl.gz` files from `2025-02-21` and later, as these contain duplicate `(track_id, mbid)` pairs and can't be applied anyway.

8) Create a new `track_mbid-update.jsonl.gz` replication file which bundles all rows `WHERE created >= '2024-02-21' OR updated >= '2024-02-21'`.

**Comment by emkey08** [comment]
However, considering the complexity of fixing the existing replication, it might actually be easier and safer to just rebuild the entire data feed now.

By the way, rebuilding the data feed should also fix https://github.com/acoustid/acoustid-server/issues/108, and that would be a nice extra.

**Comment by emkey08** [comment]
Dear @lalinsky, may I ask whether the AcoustID data files have been rebuilt by now, and/or what the current status of the data replication is now in general? Thank you very much for a short update on this!
